### PR TITLE
zv: Impl Type for std Duration and NonZero*

### LINKED ISSUE #334: zvariant: Support std::num::NonZero* types and std::time::Duration
In GitLab by @sophie-h on May 9, 2023, 01:21

### PR DESCRIPTION
**Pull request by sophie-h**
Closes #334

### CODE REVIEW

File: zvariant/src/type.rs [hunk 1]
```diff
@@ -313,6 +314,64 @@ macro_rules! map_impl {
...
 
+impl Type for Duration {
+    fn signature() -> Signature<'static> {
+        <(u64, u32)>::signature()
+    }
+}
+
+use std::num::{
+    NonZeroI16, NonZeroI32, NonZeroI64, NonZeroI8, NonZeroU16, NonZeroU32, NonZeroU64, NonZeroU8,
+};
```
Comment (zeenix) [inline zvariant/src/type.rs:325]: Either use the fully-qualified names (i-e remove this) or put this on the top with other imports.

**Review by zeenix — CHANGES_REQUESTED** [review]
Thanks. Looks good in general. Please split the `Duration` and `NonZero*` impls into 2 commits.

File: zvariant/src/type.rs [hunk 2]
```diff
@@ -313,6 +314,64 @@ macro_rules! map_impl {
...
+
+impl Type for NonZeroI8 {
+    fn signature() -> Signature<'static> {
+        i8::signature()
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (sophie-h) [reply zvariant/src/type.rs]: I promoted them to being "Basic," or is that a demotion :thinking: 

I don't see a quick way to really safe code with a macro atm

[the remark this answers is not in the document]
Comment (zeenix) [reply zvariant/src/type.rs]: > I promoted them to being "Basic," or is that a demotion thinking

:laughing: 

> I don't see a quick way to really safe code with a macro atm

Hmm? I was thinking on the lines of the same thing we do [here](https://github.com/dbus2/zbus/blob/main/zvariant/src/type.rs#L376), except taking 2 params, first the type to implement for and second the type to leach on but it's ok.

**Comment by zeenix** [comment]
Oh and can you please split the commit?

**Review by zeenix — CHANGES_REQUESTED** [review]
Waiting for split.

**Comment by sophie-h** [comment]
> Waiting for split.

done

**Review by zeenix — APPROVED** [review]

**Pull request merged**